src/daft-csv/src/local.rs

```diff
@@ -0,0 +1,924 @@
+    }
+
+    fn validate_record<'a>(&mut self, iter: &mut impl Iterator<Item = &'a u8>) -> Option<bool> {
+        // Reset state machine for each new validation attempt.
+        self.state = CsvState::FieldStart;
+        self.num_fields_seen = 1;
+        // Start running the state machine against each byte.
+        for &byte in iter {
+            let next_state = self.transition_table[self.state as usize][byte as usize];
+
+            match next_state {
+                CsvState::FieldStart => {
+                    self.num_fields_seen += 1;
+                    if self.num_fields_seen > self.num_fields {
+                        return Some(false);
+                    }
+                }
+                CsvState::RecordEnd => {
+                    return Some(self.num_fields_seen == self.num_fields);
+                }
+                CsvState::Invalid => return Some(false),
+                _ => {}
+            }
+
+            self.state = next_state;
+        }
+
+        None
+    }
+}
+
+/// Helper function that takes in a source of bytes, calls parse_csv() to extract table values from
+/// the buffer source, then returns the vector of Daft tables.
+#[allow(clippy::too_many_arguments)]
+fn collect_tables<R>(
+    has_header: bool,
+    parse_options: &CsvParseOptions,
+    byte_reader: R,
+    projection_indices: Arc<Vec<usize>>,
+    fields: Vec<Field>,
+    read_daft_fields: Arc<Vec<Arc<daft_core::datatypes::Field>>>,
+    read_schema: Arc<Schema>,
+    csv_buffer: &mut CsvBuffer,
+    include_columns: Option<Vec<String>>,
+    predicate: Option<Arc<Expr>>,
+    limit: Option<usize>,
+) -> DaftResult<Vec<Table>>
+where
+    R: std::io::Read,
+{
+    let rdr = ReaderBuilder::new()
+        .has_headers(has_header)
+        .delimiter(parse_options.delimiter)
+        .double_quote(parse_options.double_quote)
+        .quote(parse_options.quote)
+        .escape(parse_options.escape_char)
+        .comment(parse_options.comment)
+        .flexible(parse_options.allow_variable_columns)
+        .from_reader(byte_reader);
+    // The header should not count towards the limit.
+    let limit = limit.map(|limit| limit + (has_header as usize));
+    parse_csv_chunk(
+        rdr,
+        projection_indices,
+        fields,
+        read_daft_fields,
+        read_schema,
+        csv_buffer,
+        include_columns,
+        predicate,
+        limit,
+    )
+}
+
+/// Helper function that consumes a CSV reader and turns it into a vector of Daft tables.
+#[allow(clippy::too_many_arguments)]
+fn parse_csv_chunk<R>(
+    mut reader: Reader<R>,
+    projection_indices: Arc<Vec<usize>>,
+    fields: Vec<arrow2::datatypes::Field>,
+    read_daft_fields: Arc<Vec<Arc<daft_core::datatypes::Field>>>,
+    read_schema: Arc<Schema>,
+    csv_buffer: &mut CsvBuffer,
+    include_columns: Option<Vec<String>>,
+    predicate: Option<Arc<Expr>>,
+    limit: Option<usize>,
+) -> DaftResult<Vec<Table>>
```

Comment:
I think this could technically return an Iterator of Tables. so we can start emiting out smaller chunks of tables as we're still processing the Chunk. But I'm going to say it's out of scope for this PR 🥹

Reply:
Haha I did give it a shot and got a table iterator. What I didn't get to was

> start emiting out smaller chunks of tables as we're still processing the Chunk

so it kinda just sat around until someone called `.next()` on it. 2x slowdown. I think you're right, let's punt it.